Fix refresh issue

rickihastings opened the pull request:
**WIP** - Don't merge this yet, need to do more testing. But can't be bothered right now.

It's a bit hacky, I've added a session ID which lasts for a day and it tries to clear it when the game ends or when you hit the new game page etc. I've synced the redux state with `localStorage` so the game can continue where it left off. It seems to work OK for now.

There is also something weird going on when the socket reconnects in `openWebSocket`, the `session.cookieId` exists in the server but I'm not sure where it's coming from as I've not explicitly added it anywhere 🤯need to figure this out first and test it with more teams.

Might also remove the cookie and save everything in `localStorage`.

PabloDiablo commented:
Nice 🙌 . Take a look at #2. I've converted the whole thing into TS (badly - lots of `any`s 😷) and made it a single app. Debugging and refactoring should be a bit nicer now. Most files are just `.js -> .ts` so it shouldn't be too tricky to merge these together.

rickihastings commented:
Nice, looks a lot better. I noticed running the project from two terminals was pretty horrendous.

Yeah it'll be easy enough to redo my changes after you merge that.

I'll do some more testing on this tomorrow or Thursday. Never got round to it at the weekend. Not sure why, had nout else to do 😂

PabloDiablo commented:
Merged!
Something I want to do is replace the raw WebSocket with socket.io. I reckon it'd fix the connection drops we keep seeing.

rickihastings commented:
Sweet, I'll back merge it into this. Yeah I don't think the socket has any reconnect code in it, it's just raw websocket code from what I've seen. Might be wrong tho. Socket.io should definitely make it more reliable.

I need to figure out what's going on with this socket reconnect thing, and I'll test this some more.

rickihastings committed:
Re-fix refresh issue onto new branch from the HEAD

rickihastings committed:
Running prettier

rickihastings committed:
Adding socketId for debugging purposes

rickihastings committed:
Removed custom session ID logic, express-session already has this.

rickihastings commented:
Ah ok I've figured out how the session is remembering the unique socket ID when the socket reconnects, it's this `express-session` thing, I've managed to refactor out the custom session ID logic and use the session id the library provides. It seems to work OK I've checked it with a few incognito windows at the same time. Quiz master sessions can't be refreshed tho.

The only thing I'm a little worried about is that it only tries to wipe the session when it knows the game has ended, or the quiz master has left. If something went horribly wrong at the start and the game had to restart and all the clients didn't know that (message didn't successfully deliver or something) then the cookie wouldn't be wiped, which means it won't show the new game screen. Maybe it's worth adding a `/new` route or something that forcefully does it as some insurance?

rickihastings committed:
Removing todo comment

rickihastings committed:
Removing accidental change, re-ran prettier

PabloDiablo commented:
Good work! Yeah when I originally tried this I implemented a `recover-session` endpoint that checked if the session was still active. Also fyi there's a conflict because I've just pushed a change to master to support SSL sockets... finally got Let's Encrypt set up on my server.

rickihastings committed:
Merge branch 'master' into fix-refresh

rickihastings committed:
Added route to force a new team to enter incase the session has broken.

Clear local storage on refresh for quiz master so they can start new games

Removed onbeforeunload for team game as it's not required anymore

rickihastings commented:
Ok I've fixed the conflict. I've added a `/new` route to forcefully create a new team just incase something goes wrong at the start.

Btw, you've committed some questions to `src/server/data.json`, I hope they were previous ones haha.

rickihastings committed:
Use localStorage.clear instead

rickihastings committed:
Removing comment

PabloDiablo commented:
Haha they were the first week's questions 😂 
I'll merge this and deploy it to the test site 🙌

rickihastings commented:
Ah ok that's good haha.
Woohoo!
